File: tasks/study_spider/ibox/ibox_new.js

var y = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=";
function m(e) {
    for (var t = [], r = 0; r < e.length; r++) {
        var n = e.charCodeAt(r);
        n < 128 ? t.push(n) : n < 2048 ? t.push(192 | n >> 6, 128 | 63 & n) : n < 55296 || n >= 57344 ? t.push(224 | n >> 12, 128 | n >> 6 & 63, 128 | 63 & n) : (r++,
        n = 65536 + ((1023 & n) << 10 | 1023 & e.charCodeAt(r)),
        t.push(240 | n >> 18, 128 | n >> 12 & 63, 128 | n >> 6 & 63, 128 | 63 & n))
    }
    return new Uint8Array(t);
}

function get_b(o,v,g){
    b = new Uint8Array(m(JSON.stringify({
                                    method: "POST",
                                    header: o,
                                    body: v,
                                    call_id: g
                                })).buffer);
    console.log(b);
    return b;
}
var O = [
    {
        "key": "x-wx-exclude-credentials",
        "value": "unionid, cloudbase-access-token, openid"
    },
    {
        "key": "x-wx-region",
        "value": "ap-beijing"
    },
    {
        "key": "x-wx-gateway-id",
        "value": "gw-1-1g2n1gd143d56b56"
    },
    {
        "key": "host",
        "value": "api-h5-tgw.ibox.art"
    },
    {
        "key": "accept-language",
        "value": "zh-CN"
    },
    {
        "key": "ib-device-id",
        "value": "1e3fb1a68efe4ada8b1f700165d93d25"
    },
    {
        "key": "ib-trans-id",
        "value": "4ca9072c9525499684f36c2fbb22a168"
    },
    {
        "key": "x-cloudbase-phone",
        "value": ""
    },
    {
        "key": "ib-platform-type",
        "value": "web"
    },
    {
        "key": "content-type",
        "value": "application/json"
    },
    {
        "key": "user-agent",
        "value": ""
    },
    {
        "key": "x-wx-env",
        "value": "ibox-3gldlr1u1a8322d4"
    },
    {
        "key": "x-wx-call-id",
        "value": "0.8025826435162247_1673407180378"
    },
    {
        "key": "x-wx-resource-appid",
        "value": "wxb5b2c81edbd4cf69"
    },
    {
        "key": "x-wx-container-path",
        "value": "/nft-mall-web/nft/contentList"
    }
]


function s(e) {
    return e instanceof ArrayBuffer
}
function c(e) {
    this.array = e
}

function maxCompressedLength(e) {
    console.log('这是this.array')
    console.log(this.array);
    var e = e.length;
    console.log('这是e: ',)
    return 32 + e + Math.floor(e / 6)
}


function n(e, t) {
        return 506832829 * e >>> t
    }
function i(e, t) {
    return e[t] + (e[t + 1] << 8) + (e[t + 2] << 16) + (e[t + 3] << 24)
}
function o(e, t, r) {
    return e[t] === e[r] && e[t + 1] === e[r + 1] && e[t + 2] === e[r + 2] && e[t + 3] === e[r + 3]
}
function a(e, t, r, n, i) {
    return r <= 60 ? (n[i] = r - 1 << 2,
    i += 1) : r < 256 ? (n[i] = 240,
    n[i + 1] = r - 1,
    i += 2) : (n[i] = 244,
    n[i + 1] = r - 1 & 255,
    n[i + 2] = r - 1 >>> 8,
    i += 3),
    function(e, t, r, n, i) {
        var o;
        for (o = 0; o < i; o++)
            r[n + o] = e[t + o]
    }(e, t, n, i, r),
    i + r
}
function s(e, t, r, n) {
    return n < 12 && r < 2048 ? (e[t] = 1 + (n - 4 << 2) + (r >>> 8 << 5),
    e[t + 1] = 255 & r,
    t + 2) : (e[t] = 2 + (n - 1 << 2),
    e[t + 1] = 255 & r,
    e[t + 2] = r >>> 8,
    t + 3)
}
function u(e, t, r, n) {
    for (; n >= 68; )
        t = s(e, t, r, 64),
        n -= 64;
    return n > 64 && (t = s(e, t, r, 60),
    n -= 60),
    s(e, t, r, n)
}
var l = 14
  , d = new Array(l + 1);
function f(e, t, r, s, f) {
    for (var c = 1; 1 << c <= r && c <= l; )
        c += 1;
    var h = 32 - (c -= 1);
    void 0 === d[c] && (d[c] = new Uint16Array(1 << c));
    var _, p = d[c];
    for (_ = 0; _ < p.length; _++)
        p[_] = 0;
    var E, g, y, m, v, S, A, T, C, w, b = t + r, O = t, I = t, R = !0;
    if (r >= 15)
        for (E = b - 15,
        y = n(i(e, t += 1), h); R; ) {
            S = 32,
            m = t;
            do {
                if (g = y,
                A = S >>> 5,
                S += 1,
                m = (t = m) + A,
                t > E) {
                    R = !1;
                    break
                }
                y = n(i(e, m), h),
                v = O + p[g],
                p[g] = t - O
            } while (!o(e, t, v));
            if (!R)
                break;
            f = a(e, I, t - I, s, f);
            do {
                for (T = t,
                C = 4; t + C < b && e[t + C] === e[v + C]; )
                    C += 1;
                if (t += C,
                f = u(s, f, T - v, C),
                I = t,
                t >= E) {
                    R = !1;
                    break
                }
                p[n(i(e, t - 1), h)] = t - 1 - O,
                v = O + p[w = n(i(e, t), h)],
                p[w] = t - O
            } while (o(e, t, v));
            if (!R)
                break;
            y = n(i(e, t += 1), h)
        }
    return I < b && (f = a(e, I, b - I, s, f)),
    f
}
function compressToBuffer(e) {
    var t, r = this.array, n = r.length, i = 0, o = 0;
    for (o = function(e, t, r) {
        do {
            t[r] = 127 & e,
            (e >>>= 7) > 0 && (t[r] += 128),
            r += 1
        } while (e > 0);
        return r
    }(n, e, o); i < n; )
        o = f(r, i, t = Math.min(n - i, 65536), e, o),
        i += t;
    return o
}


function compress(e) {

    var t = !1
      , n = !1;
    true ? t = !0 : s(e) && (n = !0, e = new Uint8Array(e));

    var i, o, f, d = new c(e), h = maxCompressedLength(e);
    this.array = e;
    console.log('这是H: ',h)

    return t ? (i = new Uint8Array(h),
    f = compressToBuffer(i)) : n ? (i = new ArrayBuffer(h),
    o = new Uint8Array(i),
    f = d.compressToBuffer(o)) : (i = r.alloc(h),
    f = d.compressToBuffer(i)),
    i.slice(0, f)
}

// N = compress(b)
// console.log(N);

function base64ToArrayBuffer(e) {
    for (var t = function(e) {
        var t = String(e).replace(/=+$/, "")
          , r = "";
        if (t.length % 4 == 1)
            throw new Error('"atob" failed');
        for (var n = 0, i = void 0, o = void 0, a = 0; o = t.charAt(a++); ~o && (i = n % 4 ? 64 * i + o : o,
        n++ % 4) ? r += String.fromCharCode(255 & i >> (-2 * n & 6)) : 0)
            o = y.indexOf(o);
        return r
    }(e), r = t.length, n = new Uint8Array(r), i = 0; i < r; i++)
        n[i] = t.charCodeAt(i);
    return n.buffer
}

function get_x(key){
    var x = new Uint8Array(base64ToArrayBuffer(key))
    return x;
}

// x = get_x(key)
// console.log('这是x: ',x)
// console.log('这是N: ',N)
// L = D.AES_CBC.encrypt(N, x, void 0, x);

function get_params(key){
    v = 'undefined'
    g = '0.8025826435162247_1673407180378'
    b = get_b(O,v,g);
    console.log(b);
    N = compress(b);
    x = get_x(key);
    return {'x':x,'N':N}
}

// data = get_params(key)
// console.log(data)

File: tasks/study_spider/ibox/ibox_new.py
import json
import time
from Crypto.Cipher import AES
import execjs
import requests

class Ibox():

    def __init__(self):

        self.url_wxdata = 'https://web-001.cloud.servicewechat.com/wxa-qbase/jsoperatewxdata'
        self.session = requests.Session()

    def get_wxdata(self):

        time_ = str(int(time.time() * 1000))
        data = {
                "appid": "wxb5b2c81edbd4cf69",
                "data": {
                    "qbase_api_name": "tcbapi_get_service_info",
                    "qbase_req": "{\"client_random\":\"0.4085768411967976_1673322893376\",\"system\":\"\"}",
                    "qbase_options": {
                        "identityless": 'true',
                        "resourceAppid": "wxb5b2c81edbd4cf69",
                        "resourceEnv": "ibox-3gldlr1u1a8322d4",
                        "config": {
                            "database": {
                                "realtime": {
                                    "maxReconnect": 5,
                                    "reconnectInterval": 5000,
                                    "totalConnectionTimeout": 'null'
                                }
                            }
                        },
                        "appid": "wxb5b2c81edbd4cf69",
                        "env": "ibox-3gldlr1u1a8322d4"
                    },
                    "qbase_meta": {
                        "session_id": time_,
                        "sdk_version": "wx-web-sdk/WEBDOMAIN_1.0.0 (" + str(1655460325000) + ")",
                        "filter_user_info": False
                    },
                    "cli_req_id": "{}_0.23208947808707103".format(time_)
                }
            }
        print(json.dumps(data))
        headers = {
                "Accept": "*/*",
                "Accept-Encoding": "gzip, deflate, br",
                "Accept-Language": "zh-CN,zh;q=0.9",
                "Connection": "keep-alive",
                "Content-Length": "626",
                "Content-Type": "application/json",
                "Host": "web-001.cloud.servicewechat.com",
                "Origin": "https://www.ibox.art",
                "Referer": "https://www.ibox.art/",
                "sec-ch-ua": "\"Not?A_Brand\";v=\"8\", \"Chromium\";v=\"108\", \"Google Chrome\";v=\"108\"",
                "sec-ch-ua-mobile": "?0",
                "sec-ch-ua-platform": "\"Windows\"",
                "Sec-Fetch-Dest": "empty",
                "Sec-Fetch-Mode": "cors",
                "Sec-Fetch-Site": "cross-site",
                "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36"
            }
        res = self.session.post(self.url_wxdata,data=json.dumps(data),headers=headers)
        jsopera_data = res.json()['data']
        self.request_info = json.loads(jsopera_data)
        print(jsopera_data)
        print(type(jsopera_data))

    def get_detail(self):

        self.get_wxdata()
        headers = {
            "Accept": "*/*",
            "Accept-Encoding": "gzip, deflate, br",
            "Accept-Language": "zh-CN,zh;q=0.9",
            "Connection": "keep-alive",
            "Content-Length": "848",
            "Content-Type": "application/octet-stream",
            "Host": "web-001.cloud.servicewechat.com",
            "Origin": "https://www.ibox.art",
            "Referer": "https://www.ibox.art/",
            "sec-ch-ua": "\"Not?A_Brand\";v=\"8\", \"Chromium\";v=\"108\", \"Google Chrome\";v=\"108\"",
            "sec-ch-ua-mobile": "?0",
            "sec-ch-ua-platform": "\"Windows\"",
            "Sec-Fetch-Dest": "empty",
            "Sec-Fetch-Mode": "cors",
            "Sec-Fetch-Site": "cross-site",
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.0.0 Safari/537.36",
            "X-WX-COMPRESSION": "snappy",
            "X-WX-ENCRYPTION-TIMESTAMP": str(self.request_info['timestamp']),
            "X-WX-ENCRYPTION-VERSION": "2",
            "X-WX-LIB-BUILD-TS": "1655460325335",
            "X-WX-REQUEST-CONTENT-ENCODING": "JSON",
            "X-WX-RESPONSE-CONTENT-ACCEPT-ENCODING": "PB, JSON",
            "X-WX-USER-TIMEOUT": "30000"
        }

        url = self.request_info['service_url']
        key = self.request_info['key']
        data = self.get_param(key)
        res = self.session.post(url,headers=headers,data=data)
        print('这是最终返回的结果',res.text)
        # data =
#         1655460325335

    def get_param(self,key):

        time_ = str(int(time.time() * 1000))
        o = [
            {
                "key": "x-wx-exclude-credentials",
                "value": "unionid, cloudbase-access-token, openid"
            },
            {
                "key": "x-wx-region",
                "value": "ap-beijing"
            },
            {
                "key": "x-wx-gateway-id",
                "value": "gw-1-1g2n1gd143d56b56"
            },
            {
                "key": "host",
                "value": "api-h5-tgw.ibox.art"
            },
            {
                "key": "accept-language",
                "value": "zh-CN"
            },
            {
                "key": "ib-device-id",
                "value": "bdec4a5ca11b4b9a9932a7a5e1dfab99"
            },
            {
                "key": "ib-trans-id",
                "value": "b78f7d91eadb4115856727517281680a"
            },
            {
                "key": "x-cloudbase-phone",
                "value": ""
            },
            {
                "key": "ib-platform-type",
                "value": "web"
            },
            {
                "key": "content-type",
                "value": "application/json"
            },
            {
                "key": "user-agent",
                "value": ""
            },
            {
                "key": "x-wx-env",
                "value": "ibox-3gldlr1u1a8322d4"
            },
            {
                "key": "x-wx-call-id",
                "value": "0.014912083836926104_{}".format(time_)
            },
            {
                "key": "x-wx-resource-appid",
                "value": "wxb5b2c81edbd4cf69"
            },
            {
                "key": "x-wx-container-path",
                "value": "/nft-mall-web/nft/contentList"
            }
        ]
        v = 'undefined'
        time_g = str(int(time.time() * 1000))
        g = '0.8025826435162247_{}'.format(time_g)
        ctl = execjs.compile(open('./ibox_new.js','r').read())
        data = ctl.call('get_params',key)
        print('这是x: ',data['x'])
        print('这是N:', data['N'])
        data_n_new = []
        data_x_new = []
        for k,v in data['N'].items():
            data_n_new.append(v)

        for k,v in data['x'].items():
            data_x_new.append(v)

        data_n = bytes(data_n_new)
        data_x = bytes(data_x_new)
        print('这是 data_n_new: ', data_n_new)
        print('这是 data_x_new: ', data_x_new)
        print('这是 data_n: ', data_n)
        print('这是 data_x: ', data_x)
        # return data_x,data_n
        data = self.aes_encrypt(data_x,data_n)
        print('最终参数data为: ',data)
        return data
    # 需要补位，str不是16的倍数那就补足为16的倍数
    def add_to_16_byte(self,value):
        while len(value) % 16 != 0:
            value += b'\0'
        return value

    def aes_encrypt(self,key_bytes, text):
        # 增加vi向量
        aes = AES.new(key_bytes, AES.MODE_CBC, key_bytes)
        data = aes.encrypt(self.add_to_16_byte(text))
        print(data)
        return data

    def aes_(self):
        pass


if __name__ == '__main__':

    ibox = Ibox()
    # ibox.get_wxdata()
    # data = ibox.get_param()
    ibox.get_detail()